debian/drush.1
.\"                                      Hey, EMACS: -*- nroff -*-
'\" t -*- coding: UTF-8 -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH DRUSH 1 "January 11, 2011"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
drush \- shell scripting interface for Drupal
.SH SYNOPSIS
.B drush
.RI [ options ] " command" ...
.SH DESCRIPTION
.PP
\fBdrush\fP is a veritable Swiss Army knife designed to make life
easier for those of us who spend some of our working hours hacking
away at the command prompt.
.PP
\fBDrush\fP core ships with lots of useful commands for interacting with
code like modules/themes/profiles. Similarly, it runs update.php,
executes sql queries and DB migrations, and misc utilities like run
cron or clear cache.
.SH OPTIONS
.TP
.BI \-r\  path ,\ \-\-root= path
Drupal root directory to use (default: current directory)             
.TP
.BI \-l\  uri ,\ \-\-uri= uri
URI of the drupal site to use (only needed in multisite environments) 
.TP
.BI \-v ,\ \-\-verbose
Display extra information about the command.                          
.TP
.B \-q, \-\-quiet
Hide all output
.TP
.B \-d, \-\-debug
Display even more information, including internal messages.           
.TP
.B \-y, \-\-yes
Assume 'yes' as answer to all prompts                                 
.TP
.B \-s, \-\-simulate
Simulate all relevant actions (don't actually change the system)
.TP
.B \-p, \-\-pipe
Emit a compact representation of the command for scripting.           
.TP
.B \-h, \-\-help
A detailed help system, use \fBdrush help help\fP for complete help.
.TP
.B \-\-version
Show drush version.
.PP
If you get tired of typing options all the time, you can add them to
your drush.php alias or create a drushrc.php file. These provide
additional options for your drush call. They provide great flexibility
for a multi-site installation, for example. See example.drushrc.php.
.SH COMMANDS
Drush is all based around "commands" that are usually defined in
various modules. 
.TP
.B help
Print the help message. Use \-\-filter to limit command list to one
command file (e.g. \-\-filter=pm).
.TP
.B cache-clear (cc)
Clear a specific cache, or all drupal caches.
.TP
.B core-cli (cli)
Enter a new shell optimized for Drush use.
.TP
.B core-cron (cron)
Run all cron hooks.
.TP
.B core-status (st)
Provides a birds-eye view of the current Drupal installation, if any.
.TP
.B php-eval (eval)
Evaluate arbitrary php code after bootstrapping Drupal.
.TP
.B updatedb (updb)
Apply any database updates required (as with running update.php).
.TP
.B variable-get (vget)
Get a list of some or all site variables and values
.TP
.B variable-set (vset)
Set a variable.
.TP
.B watchdog show
Shows recent watchdog log messages. Optionally filter for a specific type.
.TP
.B pm-enable (en)
Enable one or more extensions (modules or themes).
.TP
.B pm-disable (dis)
Disable one or more extensions (modules or themes).
.TP
.B pm-download (dl)
Download projects from drupal.org or other sources.
.TP
.B pm-uninstall
Uninstall one or more modules.
.TP
.B pm-updatecode (upc)
Update Drupal core and contrib projects to latest recommended releases.
.TP
.B pm-update (up)
Update Drupal core and contrib projects and apply any pending database
updates (Same as pm-updatecode + updatedb).
.TP
.B sql-cli (sqlc)
Open a SQL command-line interface using Drupal's credentials.
.TP
.B sql-connect
A string for connecting to the DB.
.TP
.B sql-dump
Exports the Drupal DB as SQL using mysqldump.
.PP
This manual only lists a few of the most frequently used builtin
commands shipped with the Drush core, since other Drupal modules and
third-party "command files" may extend this list significantly. The
canonical source of information for the available commands and their
usage is the online help system.
.PP
More information on each command is also available through \fBdrush
help <command>\fP. Most commands have shortcuts indicated in
parenthesis.
.SH FILES
.IR /etc/drush/drushrc.php
.br
.I /usr/share/doc/drush/examples/example.drushrc.php
.SH SEE ALSO
.BR cvs (1),
.BR svn (1),
.BR wget (1).
.SH AUTHOR
Drush was originally developed by Arto for Drupal 4.7 (this alpha code
can still be found in the DRUPAL-4-7 branch). In May 2007, it was
partly rewritten and redesigned for Drupal 5 by frando. Since 2008,
the module has been maintained by Moshe Weitzman, Owen Barton and
Adrian Rossouw.
.PP
This manual page was written by Antoine Beaupré <anarcat@debian.org>
in June 2009 for the Debian project (but may be used by others).


debian/changelog
drush (4.5-3) unstable; urgency=low

  * fix watchfile
  * change my email address
  * remove DM field as I am now a DD

 -- Antoine Beaupré <anarcat@debian.org>  Sun, 06 Nov 2011 22:20:28 -0500

drush (4.5-2) unstable; urgency=low

  * fetch the table.inc file during the build instead of keeping a copy in
    our source repo
  * mark this package as a "single-debian-patch" to silence a lintian
    warnings - our changes are now tracked in git (Closes: #643099).
  * recommend wget *or* curl, which are both working with drush

 -- Antoine Beaupré <anarcat@koumbit.org>  Fri, 30 Sep 2011 13:13:54 -0400

drush (4.5-1) unstable; urgency=low

  * new upstream release: http://drupal.org/node/1247086
  * fix watch file to follow new naming convention
  * fix a bunch of lintian tips
  * switch to source format 3.0 (quilt)
  
 -- Antoine Beaupré <anarcat@koumbit.org>  Sun, 21 Aug 2011 15:14:15 +0100

drush (4.4-2) unstable; urgency=low

  * ship properly all the examples
  * ship properly the docs/ directory for drush topic to work
  * update to standards version 3.9.2, no change
  * do not suggest subversion/cvs but git (Closes: #628794).

 -- Antoine Beaupré <anarcat@koumbit.org>  Tue, 28 Jun 2011 14:15:06 -0400

drush (4.4-1) unstable; urgency=low

  * New upstream minor release to fix .info file.

 -- Antoine Beaupré <anarcat@koumbit.org>  Mon, 14 Mar 2011 19:22:40 -0400

drush (4.3-1) unstable; urgency=low

  * New upstream release, see http://drupal.org/node/1091572
  * Changed git repos to be the upstream repos, which have accepted that
    the Debian branch live upstream. Upstream (msonnabaum) has also
    manifested interest in maintaining the Debian/Ubuntu packages.

 -- Antoine Beaupré <anarcat@koumbit.org>  Mon, 14 Mar 2011 17:11:53 -0400

drush (4.2-1) unstable; urgency=low

  * New upstream release
  * Updated manpage (Closes: #595908)
  * Forgot to mention that the upstream release fixed the vget bug
    (Closes: #595772)

 -- Antoine Beaupré <anarcat@koumbit.org>  Thu, 20 Jan 2011 22:34:57 -0500

drush (4.1-1) unstable; urgency=low

  * New upstream release (Closes: #609362)
    This branch (4.x) is likely to live only in sid considering upstream
    is already considering a 5.x release. In general, the release cycle of
    Debian and Drush are incompatible, the latter being roughly every 6
    months. I will try to keep sid (and therefore testing) up to date with
    the upstream stable and people will be free to use 3.3 (in squeeze or
    lenny-backports) or whatever the latest stable upstream is (in sid or
    testing).

  * standards upgrade to 3.9.1 - no change
  * update table.inc to follow upstream

 -- Antoine Beaupré <anarcat@koumbit.org>  Thu, 20 Jan 2011 21:06:50 -0500

drush (3.3-1) unstable; urgency=low

  * New upstream version (Closes: #594020)
    This is the last "unclean" version of Drush 3. Further version will
    contain only security upgrades and critical bugfixes that can be
    targeted at Debian stable. As the maintainer of the Drush stable
    branch, I can commit to such a policy.

 -- Antoine Beaupré <anarcat@koumbit.org>  Fri, 27 Aug 2010 20:43:25 +0200

drush (3.1-2) unstable; urgency=low

  * fix version control urls
  * fix the drush.info issue (Closes: #588639)

 -- Antoine Beaupré <anarcat@koumbit.org>  Sun, 22 Aug 2010 19:20:02 +0200

drush (3.1-1) unstable; urgency=low

  * new upstream release, see http://drupal.org/node/833938
  * last upload mistakenly included a .diff.gz against head, this upload
    should correct that.

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 23 Jun 2010 20:13:58 -0400

drush (3.0-1) unstable; urgency=low

  * new upstream release, see http://drupal.org/node/779448 
    (Closes: #575880).
  * removed all local patches integrated upstream
  * update to standards 3.8.4, no change required

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 23 Jun 2010 19:53:13 -0400

drush (2.0-6) unstable; urgency=low

  * put this back to GPL2+ after discussions with Drupal.org people
  * update to standards 3.8.3: no change required
  * fix subversion dependencies, push php depends up to Recommends

 -- Antoine Beaupré <anarcat@koumbit.org>  Fri, 11 Sep 2009 22:47:11 +0200

drush (2.0-5) unstable; urgency=low

  * fix copyright notice to claim GPL-2 (and not GPL-2+)
  * update to standards 3.8.2: put in php section
  * backport a few fixes that have been committed to CVS after the 2.0
    release
    - #500324 dman - fails to download install profile in proper
      location. Incorrect path.
    - #493456 by guidot. three tiny errors in drush help dl Examples
      ... Also a text improvement.
    - #503328 - fix drush cron due to drush shutdown exiting. register
      exit helper in drush_shutdown. Thanks david goode
    - #533496 ZoFreX - Drush does not like being invoked from a path that
      contains spaces

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 29 Jul 2009 17:52:18 -0400

drush (2.0-5~koumbit50+1) stable; urgency=low

  * backport to lenny

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 29 Jul 2009 17:48:18 -0400

drush (2.0-4) unstable; urgency=low

  * fix paths in VCS fields
  * add DM-Upload-Allowed field since I am a Debian Maintainer
  * add VCS-* fields to debian/control file

 -- Antoine Beaupré <anarcat@koumbit.org>  Tue, 14 Jul 2009 21:21:00 -0400

drush (2.0-4~koumbit50+1) stable; urgency=low

  * backport to lenny
  * add VCS-* fields to debian/control file

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 17 Jun 2009 21:07:18 -0400

drush (2.0-3) unstable; urgency=low

  * use dh_prep instead of dh_clean to silence warnings on sid. this will
    now fail to build on lenny.
  * cleanup comments in watch and rules files
  * update to standards 3.8.1

 -- Antoine Beaupré <anarcat@koumbit.org>  Sun, 14 Jun 2009 17:10:59 -0400

drush (2.0-2) unstable; urgency=medium

  * don't use the bash wrapper because it breaks in cron

 -- Antoine Beaupré <anarcat@koumbit.org>  Sun, 14 Jun 2009 16:47:43 -0400


drush.php
#!/usr/bin/env php
<?php


/**
 * @file
 * drush is a PHP script implementing a command line shell for Drupal.
 *
 * @requires PHP CLI 4.3.0, PHP CLI 5.x, or newer.
 */

define('DRUSH_DRUPAL_BOOTSTRAP', 'includes/bootstrap.inc');

// Terminate immediately unless invoked as a command line script
if (!drush_verify_cli()) {
  die('drush.php is designed to run via the command line.');
}

exit(drush_bootstrap($GLOBALS['argc'], $GLOBALS['argv']));

/**
 * Verify that we are running PHP through the command line interface.
 *
 * This function is useful for making sure that code cannot be run via the web server,
 * such as a function that needs to write files to which the web server should not have
 * access to.
 *
 * @return
 *   A boolean value that is true when PHP is being run through the command line, 
 *   and false if being run through cgi or mod_php.
 */
function drush_verify_cli() {
  if (php_sapi_name() == 'cgi') {
    return (is_numeric($_SERVER['argc']) && $_SERVER['argc'] > 0);
  }
  
  return (php_sapi_name() == 'cli');
}

/**
 * Load drushrc files (if available) from several possible locations.
 * 
 * @param $root
 *   The Drupal root to check. 
 */
function drush_load_rc($root) {
  global $override, $args;
  
  # Specified rc file
  $configs[] = drush_get_option(array('c', 'config'), FALSE);
  # Rc file in same directory as the drush.php file
  $configs[] = dirname($_SERVER['SCRIPT_FILENAME']) . "/drushrc.php";
  # Rc file in current directory
  $configs[] = "drushrc.php";  
  # Rc file in located drupal root
  $configs[] = drush_get_option(array('r', 'root'), $root). '/drushrc.php'; 
  # Rc file in user's home directory
  $configs[] = $_SERVER['HOME'] . '/.drushrc.php';
  
  foreach ($configs as $config) {
    if (file_exists($config)) {
      define('DRUSH_CONFIG', $config);
      require_once($config);
      if (is_array($options)) {
        $args['options'] = array_merge($options, $args['options']); # Sets all the default options for drush
      }
      break;
    }
  }
}

function drush_bootstrap($argc, $argv) {
  global $args, $override, $conf;

  // Parse command line options and arguments.
  $args = drush_parse_args($argv, array('h', 'u', 'r', 'l', 'c'));
  
  // We use PWD if available because getcwd() resolves symlinks, which
  // could take us outside of the Drupal root, making it impossible to find. 
  $path = $_SERVER['PWD'];
  if (empty($path)) {
    $path = getcwd();
  }

  // Convert windows paths.
  $path = drush_convert_path($path);

  // Try and locate the Drupal root directory
  $root = _drush_locate_root($path);

  // Load .drushrc file if available. Allows you to provide defaults for options and variables.
  drush_load_rc($root);
  
  $uri = FALSE;
  // If the current directory contains a settings.php we assume that is the desired site URI.
  if (file_exists('./settings.php')) {
    // Export the following settings.php variables to the global namespace.
    global $db_url, $db_prefix, $cookie_domain, $conf, $installed_profile, $update_free_access;
    // If the settings.php has a defined path we use the URI from that.
    include_once('./settings.php');
    if (isset($base_url)) {
      $uri = $base_url;
    }
    else {
      // Alternatively we default to the name of the current directory, if it is not 'default'.
      $elements = explode('/', $path);
      $current = array_pop($elements);
      if ($current != 'default') {
        $uri = 'http://'. $current;
      }
    }
  }

  // Define basic options as constants.
  define('DRUSH_URI',         drush_get_option(array('l', 'uri'), $uri));
  define('DRUSH_VERBOSE',     drush_get_option(array('v', 'verbose'), FALSE));
  define('DRUSH_AFFIRMATIVE', drush_get_option(array('y', 'yes'), FALSE));
  define('DRUSH_SIMULATE',    drush_get_option(array('s', 'simulate'), FALSE));

  // TODO: Make use of this as soon as the external
  define('DRUSH_USER',        drush_get_option(array('u', 'user'), 0));

  // If no root is defined, we try to guess from the current directory.
  define('DRUSH_DRUPAL_ROOT',  drush_get_option(array('r', 'root'), $root));

  // If the Drupal directory can't be found, and no -r option was specified,
  // or the path specified in -r does not point to a Drupal directory,
  // we have no alternative but to give up the ghost at this point.
  // (NOTE: t() is not available yet.)
  if (!DRUSH_DRUPAL_ROOT || !is_dir(DRUSH_DRUPAL_ROOT) || !file_exists(DRUSH_DRUPAL_ROOT . '/' . DRUSH_DRUPAL_BOOTSTRAP))  {
    $dir = '';
    if (DRUSH_DRUPAL_ROOT) {
      $dir = ' in ' . DRUSH_DRUPAL_ROOT;
    }
    // Provide a helpful exit message, letting the user know where we looked
    // (if we looked at all) and a hint on how to specify the directory manually.
    $message = "E: Could not locate a Drupal installation directory$dir. Aborting.\n";
    $message .= "Hint: You can specify your Drupal installation directory with the --root\n";
    $message .= "parameter on the command line or \$options['root'] in your drushrc.php file.\n";
    die($message);
  }

  // Fake the necessary HTTP headers that Drupal needs:
  if (DRUSH_URI) {
    $drupal_base_url = parse_url(DRUSH_URI);
    $_SERVER['HTTP_HOST'] = $drupal_base_url['host'];
    $_SERVER['PHP_SELF'] = $drupal_base_url['path'].'/index.php';
  }
  else {
    $_SERVER['HTTP_HOST'] = NULL;
    $_SERVER['PHP_SELF'] = NULL;
  }
  $_SERVER['REQUEST_URI'] = $_SERVER['SCRIPT_NAME'] = $_SERVER['PHP_SELF'];
  $_SERVER['REMOTE_ADDR'] = '';
  $_SERVER['REQUEST_METHOD'] = NULL;
  $_SERVER['SERVER_SOFTWARE'] = NULL;

  // Change to Drupal root dir.
  chdir(DRUSH_DRUPAL_ROOT);
  // Bootstrap Drupal.
  _drush_bootstrap_drupal();
  /**
   * Allow the drushrc.php file to override $conf settings.
   * This is a separate variable because the $conf array gets initialized to an empty array,
   * in the drupal bootstrap process, and changes in settings.php would wipe out the drushrc.php
   * settings
   */
  if (is_array($override)) {
    $conf = array_merge($conf, $override);
  }
    
  // Login the specified user (if given).
  if (DRUSH_USER) {
    _drush_login(DRUSH_USER);
  }

  // Now we can use all of Drupal.

  if (DRUSH_SIMULATE) {
    drush_print(t('SIMULATION MODE IS ENABLED. NO ACTUAL ACTION WILL BE TAKEN. SYSTEM WILL REMAIN UNCHANGED.'));
  }

  // Dispatch the command.
  $output = drush_dispatch($args['commands']);

  // prevent a '1' at the end of the outputs
  if ($output === true) {
    $output = '';
  }

  // TODO: Terminate with the correct exit status.
  return $output;
}


/**
 * Exhaustive depth-first search to try and locate the Drupal root directory.
 * This makes it possible to run drush from a subdirectory of the drupal root.
 * 
 * @param $path
 *   The path the start the search from.
 */
function _drush_locate_root($path) {
  // Check the current path.
  if (file_exists($path . '/' . DRUSH_DRUPAL_BOOTSTRAP)) {
    return $path;
  }
  // Move up dir by dir and check each.
  while ($path = _drush_locate_root_moveup($path)) {
    if (file_exists($path . '/' . DRUSH_DRUPAL_BOOTSTRAP)) {
      return $path;
    }
  }

  return FALSE;
}

function _drush_locate_root_moveup($path) {
  if (empty($path)) {
    return FALSE;
  }
  $path = explode('/', $path);
  // Move one directory up.
  array_pop($path);
  return implode($path, '/');
}

/**
 * Bootstrap Drupal.
 */
function _drush_bootstrap_drupal() {
  require_once DRUSH_DRUPAL_BOOTSTRAP;
  if (($conf_path = conf_path()) && !file_exists("./$conf_path/settings.php")) {
    // Provide a helpful exit message, letting the user know where we looked
    // (if we looked at all) and a hint on how to specify the URI manually.
    $message = "E: Unable to load Drupal configuration from $conf_path. Aborting.\n";
    $message .= "Hint: You can specify your Drupal URI to use with the --uri\n";
    $message .= "parameter on the command line or \$options['uri'] in your drushrc.php file.\n";
    die($message);
  }

  // The bootstrap can fail silently, so we catch that in a shutdown function.
  register_shutdown_function('drush_shutdown');
  drupal_bootstrap(DRUPAL_BOOTSTRAP_FULL);
  if (module_exists('drush')) {
    require_once drupal_get_path('module', 'drush') . '/drush.inc';
  }
  else {
    $message = "E: You must enable the Drush module for the site you want to use.\n";
    $message .= "Hint: Drush was looking in the site '$conf_path'. You can select another site\n";
    $message .= "with Drush enabled by specifying the Drupal URI to use with the --uri\n";
    $message .= "parameter on the command line or \$options['uri'] in your drushrc.php file.\n";
    die($message);
  }
}

function drush_shutdown() {
  if (!function_exists('drupal_set_content')) {
    $message = "E: Drush was not able to start (bootstrap) Drupal.\n";
    $message .= "Hint: This error often occurs when Drush is trying to bootstrap a site\n";
    $message .= "that has not been installed or does not have a configured \$db_url.\n";
    $message .= "Drush was looking in the site '$conf_path'. You can select another site\n";
    $message .= "with a working database setup by specifying the URI to use with the --uri\n";
    $message .= "parameter on the command line or \$options['uri'] in your drushrc.php file.\n";
    die($message);
  }
}

/**
 * Log a certain user in.
 */
function _drush_login($drush_user) {
  global $user;
  $user = user_load(is_numeric($drush_user) ? array('uid' => $drush_user) : array('name' => $drush_user));

  if (empty($user)) {
    if (is_numeric($drush_user)) {
      drush_die(t('Could not login with user ID #%user.', array('%user' => $drush_user)));
    }
    else {
      drush_die(t('Could not login with user account `%user\'.', array('%user' => $drush_user)));
    }
    return FALSE;
  }

  return TRUE;
}

/**
 * Parse console arguments.
 *
 * @param $args
 *   The console argument array (usually $argv)
 * @param $arg_opts
 *   An array of options that are followed by an argument.
 *   e.g. shell.php -u admin -v --> $arg_opts = array('u')
 * @param $default_options
 * @return
 *   A associative array:
 *   $return['commands'] ia a numeric array of all commands,
 *   $return['options'] contains the options. The option keys
 *   are always noted without - or -- and are set to TRUE if they were
 *   invoked, to the argument if followed by an argument, and if not present
 *   to their default value or FALSE if no default value was specified.
 */
function drush_parse_args($args = array(), $arg_opts = array(), $default_options = array()) {
  $options = $default_options;
  $commands = array();

  for ($i = 1; $i < count($args); $i++) {
    $opt = $args[$i];
    // Is the arg an option (starting with '-')?
    if ($opt{0} == "-" && strlen($opt) != 1) {
      // Do we have multiple options behind one '-'?
      if (strlen($opt) > 2 && $opt{1} != "-") {
        // Each char becomes a key of its own.
        for ($j = 1; $j < strlen($opt); $j++) {
          $options[substr($opt, $j, 1)] = true;
        }
      }
      // Do we have a longopt (starting with '--')?
      elseif ($opt{1} == "-") {
        if ($pos = strpos($opt, '=')) {
          $options[substr($opt, 2, $pos - 2)] = substr($opt, $pos + 1);
        }
        else {
          $options[substr($opt, 2)] = true;
        }
      }
      else {
        $opt = substr($opt, 1);
        // Check if the current opt is in $arg_opts (= has to be followed by an argument).
        if ((in_array($opt, $arg_opts))) {
          if (($args[$i+1] == NULL) || ($args[$i+1] == "") || ($args[$i + 1]{0} == "-")) {
            exit("Invalid input: -$opt needs to be followed by an argument.");
          }
          $options[$opt] = $args[$i + 1];
          $i++;
        }
        else {
          $options[$opt] = true;
        }
      }
    }
    // If it's not an option, it's a command.
    else {
      $commands[] = $opt;
    }
  }
  return array('options' => $options, 'commands' => $commands);
}

/**
 * Get the value for an option.
 *
 * If the first argument is an array, then it checks wether one of the options
 * exists and return the value of the first one found. Useful for allowing both
 * -h and --host-name
 *
 */
function drush_get_option($option, $default = NULL) {
  $options = $GLOBALS['args']['options'];
  if (is_array($option)) {
    foreach ($option as $current) {
      if (array_key_exists($current, $options)) {
        return $options[$current];
      }
    }
    return $default;
  }

  if (!array_key_exists($option, $options)) {
    return $default;
  }
  else {
    return $options[$option];
  }
}

/**
 * Converts a Windows path (dir1\dir2\dir3) into a Unix path (dir1/dir2/dir3).
 */
function drush_convert_path($path) {
  return str_replace('\\','/', $path);
}

